### DISCUSSION TITLE
support signal SIGHUB reload modules config

### DISCUSSION THREAD

**Discussion by kwanhur** [bfenetworks/bfe#892] [Ideas（新功能的建议和讨论）]
<sub>Sent from <a href="https://apps.apple.com/cn/app/id1314212521">PPHub</a></sub>

**Comment by songoow** [comment]
说一下我的理解，如果存在理解上的错误，请指正。目前这个想法是想通过 SIGHUB来重新加载 bfe/conf/.../  ....data 这样的配置文件吗，目前的重载实现的方式是通过conf-agent里面for循环sleep 一定的时间来发送http 请求来对比是否配置文件已经被更新了，如果更新了还是通过http 请求发送给bfe 来重新加载对应配置。目前在bfe 上面还没有对配置文件的监控，如果使用sighub 的话，当有一个配置修改了，无法定位到是那个配置修改，需要对比全部的配置文件来定位，或者直接对所有对应的配置文件都重新加载一遍？所以我想这里是不是可以采取这样的方式，https://github.com/spf13/viper 通过viper 来在每个配置文件初次加载的时候，使用viper 来监控对应的配置文件，当配置文件有所改动的时候，可以直接监控到变化，直接就更新了对应的配置文件。如果viper 的配置动态更新机制适合这边的场景，我这边可以在bfe的基础上来尝试实现一版。

**Comment by kwanhur** [comment]
Thx for response :-)

This idea comes from `Nginx`, when BFE deploy as a raw process service on host, after configuration's files modified, always command `systemctl reload bfe` or `kill -HUB $(pidof bfe)` to trigger bfe update inner configurations gracefully.

Support `conf/bfe.conf` and `conf/mod_*/*` files will be the best.

**Comment by songoow** [comment]
https://github.com/bfenetworks/bfe/pull/887#issuecomment-967034789  

Some Results of discussion:

1 : bfe  more suitable for using  http trigger  reload update configurations , in  conf-agent  provided  temporary conf File  which prepare check the conf file  will  more  security。

2 : conf -agent  pull conf from  api-server , which is easy  to implement than api-server push to conf-agent.

3: Configuration center will make the bfe more complex to build，the current approach is simpler.

**Comment by songoow** [comment]
My ppt of discussion：
![image](https://user-images.githubusercontent.com/22676124/141609820-112089d4-09d5-4bf5-9570-3d52c5449bbe.png)
![image](https://user-images.githubusercontent.com/22676124/141609829-a1c916d5-fc99-4593-a72c-a9b1d59ffa8e.png)
![image](https://user-images.githubusercontent.com/22676124/141609841-60958ce5-6c7a-4ab9-a69e-8b93d65df83f.png)
![image](https://user-images.githubusercontent.com/22676124/141609849-bf9a2272-8d53-4170-95a8-815efe1fcc9a.png)
![image](https://user-images.githubusercontent.com/22676124/141609858-9b524807-39a2-4e23-a0df-9a09672083a8.png)
![image](https://user-images.githubusercontent.com/22676124/141609868-79ad4814-73a7-4740-861b-f701297603cb.png)
![image](https://user-images.githubusercontent.com/22676124/141609875-046c1e78-08c5-40b2-bb65-1bbe1926c225.png)
![image](https://user-images.githubusercontent.com/22676124/141609879-307b58d9-6250-4219-96f1-f40943f16b37.png)

**Comment by kwanhur** [comment]
Config center solution like `etcd` `consul` `nacos` sounds the better between `conf-agent` and `bfe`.

When `bfe` just run locally, then it needs to focus on configuration files.

From `viper`'s description, it has these abilities. `viper` maybe a candidate solution.
